Util/Lexer.pas

```diff
@@ -0,0 +1,226 @@
+unit Lexer;
+
+interface
+
+uses
+  Generics.Defaults,
+  Generics.Collections,   // for TList<>
+  SysUtils,               // Exception
+  PerlRegEx;      // for regular expression support (TPerlRegEx, TPerlRegExList)
+
+type
+  {
+    Lexer rules are defined as a pair, consisting of a RegEx string and a
+    corresponding type. An array of TTokenRuleDef records will be passed
+    into TLexer<T>.Create.
+
+    Example:
+      LexerRules: array[0..2] of TTokenRuleDef = (
+        (R: '[A-Z]+';               T: Ord(ttAlpha)),
+        (R: '\d+';                  T: Ord(ttNumeric)),
+        (R: '[A-Z][A-Z\d]*';        T: Ord(ttAlphaNumeric))
```

Comment:
Unless I miss something it *must* start with a letter, so it's not "AlphaNumeric" per se. "1A" won't match for example. I understand it's only examples, but it's a bit confusing.

Reply:
Good point. I initially started thinking the ttAlphaNumeric and ttNumericAlpha tokens would be useful, but they weren't. In the end for ARRL SS Contest, I converged on three token types: ttAlpha, ttDigits (numeric), and ttCallsign. I will update the above example to remove the ttAlphaNumeric. I will update this example later today or tonight.

In the few days, I'll post the ARRL SS Contest for review. It will show the tables used for SS and will hopefully be more clear. Parsing has never been my strength, so my examples may not be the best to provide.

Thank you for the feedback.

@f6fvy

Reply:
I updated the comments and pushed an updated version.